From 00374fab3e4b9ecb878e86f03765aa8fb39059b8 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:51:29 -0400
Subject: [fs] nfsd:Avoid NULL reference or double free in nfsd4_fslocs_free()

Message-id: <1403560314-32525-148-git-send-email-bfields@redhat.com>
Patchwork-id: 84525
O-Subject: [Patch RHEL7 147/172] NFS4: Avoid NULL reference or double free in nfsd4_fslocs_free()
Bugzilla: 1109733 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

If fsloc_parse() failed at kzalloc(), fs/nfsd/export.c
 411
 412         fsloc->locations = kzalloc(fsloc->locations_count
 413                         * sizeof(struct nfsd4_fs_location), GFP_KERNEL);
 414         if (!fsloc->locations)
 415                 return -ENOMEM;

svc_export_parse() will call nfsd4_fslocs_free() with fsloc->locations = NULL,
so that, "kfree(fsloc->locations[i].path);" will cause a crash.

If fsloc_parse() failed after that, fsloc_parse() will call nfsd4_fslocs_free(),
and svc_export_parse() will call it again, so that, a double free is caused.

This patch checks the fsloc->locations, and set to NULL after it be freed.

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: a1f05514b016aeaed638dbf677f443af7e7bde4f
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/export.c b/fs/nfsd/export.c
index 8513c59..9a41d3d 100644
--- a/fs/nfsd/export.c
+++ b/fs/nfsd/export.c
@@ -295,13 +295,19 @@ svc_expkey_update(struct cache_detail *cd, struct svc_expkey *new,
 
 static void nfsd4_fslocs_free(struct nfsd4_fs_locations *fsloc)
 {
+ struct nfsd4_fs_location *locations = fsloc->locations;
  int i;
 
+ if (!locations)
+  return;
+
  for (i = 0; i < fsloc->locations_count; i++) {
-  kfree(fsloc->locations[i].path);
-  kfree(fsloc->locations[i].hosts);
+  kfree(locations[i].path);
+  kfree(locations[i].hosts);
  }
- kfree(fsloc->locations);
+
+ kfree(locations);
+ fsloc->locations = NULL;
 }
 
 static void svc_export_put(struct kref *ref)
-- 
1.7.1